### ISSUE TITLE
Drop test flaky

### ISSUE THREAD

**Issue by lucas-clemente** [lucas-clemente/quic-go#555]
```
[BeforeEach] [Top Level]
  /home/travis/gopath/src/github.com/lucas-clemente/quic-go/integrationtests/integrationtests_suite_test.go:83
[BeforeEach] Drop Proxy
  /home/travis/gopath/src/github.com/lucas-clemente/quic-go/integrationtests/drop_test.go:22
[It] gets a file when many outgoing packets are dropped
  /home/travis/gopath/src/github.com/lucas-clemente/quic-go/integrationtests/drop_test.go:68
Connected to 127.0.0.1:12345
Request:
headers:
{
  :method GET
  :scheme https
  :authority quic.clemente.io
  :path /data
}
body: 
Response:
headers: 
{
  :status 200
}
body: 
trailers: {}
Request succeeded (200).
[AfterEach] Drop Proxy
  /home/travis/gopath/src/github.com/lucas-clemente/quic-go/integrationtests/drop_test.go:52
[AfterEach] [Top Level]
  /home/travis/gopath/src/github.com/lucas-clemente/quic-go/integrationtests/integrationtests_suite_test.go:94
• Failure [4.412 seconds]
Drop Proxy with quic version 35 dropping every 4th packet after the crypto handshake [It] gets a file when many outgoing packets are dropped 
/home/travis/gopath/src/github.com/lucas-clemente/quic-go/integrationtests/drop_test.go:68
  Expected error:
      <*qerr.QuicError | 0x20c22a90>: {
          ErrorCode: 25,
          ErrorMessage: "No recent network activity.",
      }
      NetworkIdleTimeout: No recent network activity.
  not to have occurred
  /home/travis/gopath/src/github.com/lucas-clemente/quic-go/integrationtests/integrationtests_suite_test.go:110
  Full Stack Trace
  	/home/travis/gopath/src/github.com/onsi/gomega/internal/assertion/assertion.go:69 +0x164
  github.com/onsi/gomega/internal/assertion.(*Assertion).NotTo(0x20c08500, 0x85f33c0, 0x865fa24, 0x0, 0x0, 0x0, 0x20c08500)
  	/home/travis/gopath/src/github.com/onsi/gomega/internal/assertion/assertion.go:43 +0x85
  github.com/lucas-clemente/quic-go/integrationtests.setupHTTPHandlers.func2(0x85f31e0, 0x20a6ffb0, 0x18968880)
  	/home/travis/gopath/src/github.com/lucas-clemente/quic-go/integrationtests/integrationtests_suite_test.go:110 +0x1b4
  net/http.HandlerFunc.ServeHTTP(0x8451e8c, 0x85f31e0, 0x20a6ffb0, 0x18968880)
  	/home/travis/.gimme/versions/go1.8.1.linux.amd64/src/net/http/server.go:1942 +0x34
  net/http.(*ServeMux).ServeHTTP(0x864e460, 0x85f31e0, 0x20a6ffb0, 0x18968880)
  	/home/travis/.gimme/versions/go1.8.1.linux.amd64/src/net/http/server.go:2238 +0xf0
  github.com/lucas-clemente/quic-go/h2quic.(*Server).handleRequest.func1.1(0x1892dfcb, 0x85f04a0, 0x864e460, 0x20a6ffb0, 0x18968880)
  	/home/travis/gopath/src/github.com/lucas-clemente/quic-go/h2quic/server.go:207 +0x61
  github.com/lucas-clemente/quic-go/h2quic.(*Server).handleRequest.func1(0x18ad10c0, 0x20a6ffb0, 0x18968880, 0x20a99b01, 0x20a72f20, 0xf7682b70, 0x18d49b30)
  	/home/travis/gopath/src/github.com/lucas-clemente/quic-go/h2quic/server.go:208 +0x56
  created by github.com/lucas-clemente/quic-go/h2quic.(*Server).handleRequest
  	/home/travis/gopath/src/github.com/lucas-clemente/quic-go/h2quic/server.go:224 +0x427
  

```

**Comment by marten-seemann** [comment]
I managed to get a log: 
[log.txt](https://github.com/lucas-clemente/quic-go/files/948707/log.txt)

**Comment by marten-seemann** [comment]
Using the new QUIC proxy (#560), I'm getting the following:
[log.txt](https://github.com/lucas-clemente/quic-go/files/949157/log.txt)
Looks like a completely different bug, probably something with loss recovery and congestion.

**Comment by lucas-clemente** [comment]
Just to be clear, is the issue from the first log fixed? If not, I'll open a separate bug for the new issue.

**Comment by marten-seemann** [comment]
The first one is related to #544. No need to open a new issue for that.

**Comment by lucas-clemente** [comment]
ok :)

**Comment by lucas-clemente** [comment]
If we loose a lot of packets in a row, we decrease the congestion window faster than we're reducing the bytes in flight. This is correct behavior, but means that we basically have to wait for all packets to RTO. As our RTO increases exponentially (capped to 60s), this could take an extremely long time. We also don't send ACKs anymore (#576), which is why the PINGs don't work as expected, and the connection finally aborts.

We should definitely keep sending ACKs, which would allow us to make progress. Also, I think this would be mitigated by TLPs (#497), as they should reduce the likelihood for RTOs.

I'm not sure whether that's the full solution, we should probably raise this with gQUIC people.

**Comment by f3z0** [comment]
After applying a penalty to congestion window size and retransmission timeout what if we implement a short cool-down period before any further penalization can occur. I think this makes sense because it gives the network an opportunity to adjust to the newest configuration, this way we avoid run-away over-correcting when multiple packets are lost in a row. Would this be within the bounds of quic-protocol specifications, I'm guessing not.

**Comment by lucas-clemente** [comment]
There's two fixes to this: First, we shouldn't immediately collapse the cwnd, and instead wait until we're sure we actually lost the packet (i.e. the next ACK of a higher packet number). This is also useful to detect spurious ACKs.

Second, we should send retransmissions for RTOs (possibly for other losses?) if we are cwnd-limited. We can keep a counter for this, bi-increment it on RTO and decrement it for each retransmission.

I'll look into fixing the second issue first, and will sync about the details with Jana at the IETF interim this week.

**Comment by marten-seemann** [comment]
Here another log: https://pastebin.com/6mknmdmJ

**Comment by lucas-clemente** [comment]
I don't see a CONNECTION_CLOSE in this log – do you remember what the error condition was and for which test exactly you were collecting the log?

**Comment by marten-seemann** [comment]
Not 100% sure, but I think it was a timeout.


On June 11, 2017 at 11:55:50, Lucas Clemente (notifications@github.com)
wrote:

I don't see a CONNECTION_CLOSE in this log – do you remember what the error
condition was and for which test exactly you were collecting the log?

—
You are receiving this because you were assigned.
Reply to this email directly, view it on GitHub
<https://github.com/lucas-clemente/quic-go/issues/555#issuecomment-307617933>,
or mute the thread

**Comment by marten-seemann** [comment]
(Probably) fixed by #1053, #1057, #1058.
